### PR TITLE
rewrite shell scripts to Python

### PR DESCRIPTION
**Pull request by vladak**
This changes the implementation of all frequently used scripts to Python. The `OpenGrok` shell script became thin layer on top of executing `java` with `opengrok.jar` (same goes for `Groups` and `ConfigMerge` scripts). It is now called `indexer.py`. It attempts to find `java` binary however user has to supply the path to `opengrok.jar` himself. All the scripts now live in the `tools` directory. This also finally changes the last pieces of script to use `requests` Python module to make RESTful API calls. There is a new script `deploy.py` that facilitates the same as previously done by the `OpenGrok deploy` subcommand, except it does not perform detection of the application server anymore (it was too fragile anyway).

A small transition guide is here: https://gist.github.com/vladak/793307491ac81b01aa40b224be698fbc

Collateral changes:
  - `Command` now supports setting resource limits
  - `Command` argument substitution changed so that it is performed anywhere in any argument
  - `sync.py` configuration file has changed to support said resource limits and environment variables settings

Once this is merged in, I will update:
  - README
  - https://github.com/oracle/opengrok/wiki/How-to-install-OpenGrok
  - https://github.com/oracle/opengrok/wiki/Repository-synchronization

### CODE REVIEW

File: tools/command.py [hunk 1]
```diff
@@ -264,28 +271,51 @@ def fill_arg(self, args_append=None, args_subst=None):
...
+        raise NotImplementedError("unknown resource")
+
+    def set_resource_limit(self, name, value):
+        self.logger.debug("Setting resource {} to {}"
+                          .format(name, value))
+        resource.setrlimit(self.get_resource(name), (value, value))
+
+    def set_resource_limits(self, limits):
+        self.logger.debug("Setting resource limits")
+        for name in limits.keys():
```
Comment (tehe) [inline tools/command.py]: Assuming this is Python 3, one would better write the code as follows:
```python
for name, value in limits.items():
    self.set_resource_limit(name, value)
```

File: tools/deploy.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        if ret is None or ret != 0:
+            raise Exception("Cannot re-compress war file {}, command '{}' "
+                            "ended with {}".format(targetWar, cmd, ret))
+
+        # Need to switch back to original working directory, so that the
+        # temporary directory can be deleted.
+        os.chdir(origWorkingDir)
+
+
+def deployWar(logger, sourceWar, targetWar, configFile=None):
```
Comment (tulinkry) [inline tools/deploy.py]: `deploy_war`

File: scripts/before_install [hunk 3]
```diff
@@ -1,10 +1,14 @@
 #!/bin/bash
 
-if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then sudo apt-get update -qq; fi
 if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
-	sudo apt-get install -qq exuberant-ctags cvs git mercurial cssc bzr subversion monotone rcs pep8;
+	sudo apt-get update -qq
+	sudo apt-get install -qq exuberant-ctags cvs git mercurial cssc bzr subversion monotone rcs python3 python3-pip pep8;
```
[the remark this answers is not in the document]
Comment (vladak) [reply scripts/before_install]: https://www.quora.com/What-is-the-difference-between-PEP8-and-Flake8-Python-coding-standards says:

> flake8 is “the wrapper which verifies pep8, pyflakes and circular complexity “

File: tools/indexer.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    logger = logging.getLogger(os.path.basename(sys.argv[0]))
+
+    #
+    # Since it is not possible to tell what kind of action is performed,
+    # always check ctags presence unless told not to.
+    #
+    if not args.no_ctags_check and not FindCtags(logger):
+        logger.warning("Unable to determine Exuberant/Universal CTags command")
+
+    indexer = Indexer(args.options, logger=logger, java=args.java,
```
[the remark this answers is not in the document]
Comment (vladak) [reply tools/indexer.py]: This is exactly what I wanted to avoid - parsing OpenGrok options.

**Comment by vladak** [comment]
I will just merge this now, produce new RC and @Orviss you can try it on the demo instance. Any bugs found can be addressed separately. Also, I will setup periodic mirroring/indexing on my mac to have additional coverage.

**Pull request merged**